bindings/c/test/mako/mako.c

```diff
@@ -478,172 +426,224 @@ int run_transaction(FDBTransaction *transaction, mako_args_t *args,
-                  randstrlen + 1); /* make it (almost) unique */
-          randstr(valstr, args->value_length + 1);
-          for (rangei = 0; rangei < args->txnspec.ops[i][OP_RANGE]; rangei++) {
-            sprintf(keystr + KEYPREFIXLEN + randstrlen, "%0.*d",
-                    digits(args->txnspec.ops[i][OP_RANGE]), rangei);
-            if (rangei == 0) {
-              strcpy(keystr2, keystr);
-              keystr2[strlen(keystr)] = '\0';
-            }
-            rc = run_op_insert(transaction, keystr, valstr);
-            if (rc != 0) {
-              /* rollback not necessary, transaction will be reset */
-              break;
-            }
-          }
-          /* commit inserts so mutation goes to storage */
-          if (commit_transaction(transaction, stats) == 0) {
-            clock_gettime(CLOCK_MONOTONIC, &timer_per_xact_end);
-            update_op_stats(&timer_per_xact_start, &timer_per_xact_end,
-                            OP_COMMIT, stats);
-          }
-          fdb_transaction_reset(transaction);
-          rc = run_op_clearrange(transaction, keystr2, keystr);
-          docommit = 1;
-          break;
-        default:
-          fprintf(stderr, "ERROR: Unknown Operation %d\n", i);
-          break;
-        }
-
-        if (stats->xacts % args->sampling == 0) {
-          clock_gettime(CLOCK_MONOTONIC, &timer_end);
-          if (rc == 0) {
-            /* per op latency */
-            update_op_stats(&timer_start, &timer_end, i, stats);
-          }
-        }
+	if (stats->xacts % args->sampling == 0) {
+	  clock_gettime(CLOCK_MONOTONIC, &timer_end);
+	  if (rc == FDB_SUCCESS) {
+	    /* per op latency, record successful transactions */
+	    update_op_stats(&timer_start, &timer_end, i, stats);
+	  }
+	}
 
-        /* check rc */
-        if (rc != 0) {
-          stats->errors[i]++;
-        } else {
-          stats->ops[i]++;
-        }
+	/* check rc */
+	if (rc != FDB_SUCCESS) {
+	  stats->errors[i]++;
+	  if (rc == FDB_ERROR_ABORT) {
+	    goto abortTxn;
+	  }
+	  /* no need to run the rest, immediately retry */
+	  goto retryTxn;
+	} else {
+	  /* success */
+	  stats->ops[i]++;
+	}
       }
     }
   }
+
+  /* commit only successful transaction */
   if (docommit | args->commit_get) {
-    if (commit_transaction(transaction, stats) == 0) {
+    rc = commit_transaction(transaction);
+    if (rc == FDB_SUCCESS) {
+      /* success */
+      stats->ops[OP_COMMIT]++;
       clock_gettime(CLOCK_MONOTONIC, &timer_per_xact_end);
-      update_op_stats(&timer_per_xact_start, &timer_per_xact_end, OP_COMMIT,
-                      stats);
+      update_op_stats(&timer_per_xact_start, &timer_per_xact_end,
+		      OP_COMMIT, stats);
+    } else {
+      if (rc == FDB_ERROR_CONFLICT) {
+	stats->conflicts++;
+	goto retryTxn;
+      } else {
+	stats->errors[OP_COMMIT]++;
+	if (rc == FDB_ERROR_RETRY) {
+	  goto retryTxn;
+	} else {
+	  goto abortTxn;
+	}
+      }
     }
   }
+
   stats->xacts++;
 
+ abortTxn:
```

Comment:
Do you need to reset the transaction here? are you going to reuse the transaction object after you return?

Reply:
It's reset at the end of `run_one_transaction()` function.  But you've got a good point.  It can be much easier to read if it just resets within `fdb_wait_and_handle_error` macro.  Will change that.

Reply:
Ah I see now - you use one transaction per thread - I guess this makes sense. I just got a bit confused by the contract you're using (`run_one_transaction` should always return with a reset transaction).

Would it be cleaner to call `fdb_transaction_reset` at the very beginning of this function? This would safeguard against bugs where one returns out of this function with a dirty transaction state...

Reply:
That's a very good suggestion!  Let me do that.